### PR TITLE
Move adaptive sampling processor

### PR DESCRIPTION
**Pull request by black-adder**
Signed-off-by: Won Jun Jang <wjang@uber.com>

This PR is part 2 of moving adaptive sampling (#365) over to OSS. This is moving over processor and related components.

### CODE REVIEW

File: plugin/sampling/strategystore/adaptive/cache.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adaptive
+
+// samplingCacheEntry keeps track of the probability and whether a service-operation is using adaptive sampling
+type samplingCacheEntry struct {
+	probability    float64
+	usingAdapative bool
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/cache.go]: s/usingAdapative/usingAdaptive

File: plugin/sampling/strategystore/adaptive/cache.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adaptive
+
+// samplingCacheEntry keeps track of the probability and whether a service-operation is using adaptive sampling
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/cache.go]: "whether a service-operation is using adaptive sampling" - in what sense, configured to use or observed to use?

File: plugin/sampling/strategystore/adaptive/cache.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+type samplingCacheEntry struct {
+	probability    float64
+	usingAdapative bool
+}
+
+type samplingCache map[string]map[string]*samplingCacheEntry
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/cache.go]: `// nested map: service -> operation -> cache entry`

File: plugin/sampling/strategystore/adaptive/factory.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling/strategystore"
+)
+
+// Factory implements strategystore.Factory for an adaptive strategy store.
+type Factory struct {
+	options        *Options
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/factory.go]: don't need a pointer (faster access)

File: plugin/sampling/strategystore/adaptive/options.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	CalculationInterval time.Duration
+
+	// LookbackInterval is the total amount of throughput data used to calculate probabilities.
+	LookbackInterval time.Duration
+
+	// Delay is the amount of time to delay probability generation by, ie. if the calculationInterval
+	// is 1 minute, the number of buckets is 10, and the delay is 2 minutes, then at one time
+	// we'll have [now()-12,now()-2] range of throughput data in memory to base the calculations
+	// off of.
+	Delay time.Duration
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/options.go]: could use a bit more explanation why it's needed

File: plugin/sampling/strategystore/adaptive/options.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	// Increase this to reduce the amount of fluctuation in the probability calculation.
+	QPSEquivalenceThreshold float64
+
+	// LookbackQPSCount determines how many previous operation QPS are used in calculating the weighted QPS,
+	// ie. if LookbackQPSCount is 1, the only the most recent QPS will be used in calculating the weighted QPS.
+	LookbackQPSCount int
+
+	// CalculationInterval determines the interval each bucket represents, ie. if an interval is
+	// 1 minute, the bucket will contain 1 minute of throughput data for all services.
+	CalculationInterval time.Duration
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/options.go]: so more like recording or aggregation interval, rather than calculation? It doesn't sound like it represents any calculation

File: plugin/sampling/strategystore/adaptive/options.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	// is 1 minute, the number of buckets is 10, and the delay is 2 minutes, then at one time
+	// we'll have [now()-12,now()-2] range of throughput data in memory to base the calculations
+	// off of.
+	Delay time.Duration
+
+	// DefaultSamplingProbability is the initial sampling probability for all new operations.
+	DefaultSamplingProbability float64
+
+	// MinSamplingProbability is the minimum sampling probability for all operations. ie. the calculated sampling
+	// probability will be bound [MinSamplingProbability, 1.0]
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/options.go]: will be bound to min(MinSamplingProbability, 1.0)

File: plugin/sampling/strategystore/adaptive/options.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	FollowerLeaseRefreshInterval time.Duration
+}
+
+// AddFlags adds flags for Options
+func AddFlags(flagSet *flag.FlagSet) {
+	flagSet.Float64(targetQPS, defaultTargetQPS,
+		"The target number of sampled traces for all operations.",
+	)
+	flagSet.Float64(equivalenceThreshold, defaultEquivalenceThreshold,
+		"The acceptable amount of deviation for the operation QPS from the `targetQPS`. Increase this to reduce the amount of fluctuation in the probability calculation.",
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/options.go]: acceptable for what? If the actual delta is smaller, then ... ?  No update to the probability?

File: plugin/sampling/strategystore/adaptive/options.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	// LookbackQPSCount determines how many previous operation QPS are used in calculating the weighted QPS,
+	// ie. if LookbackQPSCount is 1, the only the most recent QPS will be used in calculating the weighted QPS.
+	LookbackQPSCount int
+
+	// CalculationInterval determines the interval each bucket represents, ie. if an interval is
+	// 1 minute, the bucket will contain 1 minute of throughput data for all services.
+	CalculationInterval time.Duration
+
+	// LookbackInterval is the total amount of throughput data used to calculate probabilities.
+	LookbackInterval time.Duration
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/options.go]: not sure I understand how this is different from LookbackQPSCount. 

Isn't `LookbackInterval == LookbackQPSCount * CalculationInterval` ?

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+
+const (
+	maxSamplingProbability = 1.0
+
+	getThroughputErrMsg = "failed to get throughput from storage"
+
+	defaultFollowerProbabilityInterval = 20 * time.Second
+
+	// The number of past entries for samplingCache the leader keeps in memory
+	serviceCacheSize = 25
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: can it be < lookbackQPSCount?

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+)
+
+var (
+	errIntervals        = errors.New("calculationInterval must be less than LookbackInterval")
+	errNonZeroIntervals = errors.New("calculationInterval and LookbackInterval must be greater than 0")
+	errLookbackQPSCount = errors.New("lookbackQPSCount cannot be less than 1")
+)
+
+type serviceOperationThroughput map[string]map[string]*model.Throughput
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: `// nested map: service -> operation  -> cache entry`

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+	p.RUnlock()
+
+	usingAdaptiveSampling := p.usingAdaptiveSampling(oldProbability, service, operation, latestThroughput)
+	p.serviceCache[0].Set(service, operation, &samplingCacheEntry{
+		probability:    oldProbability,
+		usingAdapative: usingAdaptiveSampling,
+	})
+
+	// Short circuit if the qps is close enough to targetQPS or if the service isn't using
+	// adaptive sampling.
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: using or appears to be using adaptive sampling?

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+	if math.Abs(qps-p.TargetQPS) < p.QPSEquivalenceThreshold || !usingAdaptiveSampling {
+		return oldProbability
+	}
+	var newProbability float64
+	if floatEquals(qps, 0) {
+		// Edge case, we double the sampling probability if the QPS is 0 so that we force the service
+		// to at least sample one span probabilistically
+		newProbability = oldProbability * 2.0
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: I thought we had protection against quick increase. x2 seems pretty sudden.

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+		newProbability = p.probabilityCalculator.Calculate(p.TargetQPS, qps, oldProbability)
+	}
+	return math.Min(maxSamplingProbability, math.Max(p.MinSamplingProbability, newProbability))
+}
+
+func combineProbabilities(p1 map[string]struct{}, p2 map[string]struct{}) map[string]struct{} {
+	for probability := range p2 {
+		p1[probability] = struct{}{}
+	}
+	return p1
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: doesn't seem to combine anything, only ensure keys(p1) are a superset of keys(p2)

also, calling a key "probability" looks weird, how is string a probability?

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+		f := truncateFloat(probability)
+		_, ok = opThroughput.Probabilities[f]
+		return ok
+	}
+	// By this point, we know that there's no recorded throughput for this operation for this round
+	// of calculation. Check the previous bucket to see if this operation was using adaptive sampling
+	// before.
+	if len(p.serviceCache) > 1 {
+		if e := p.serviceCache[1].Get(service, operation); e != nil {
+			return e.usingAdapative && e.probability != p.DefaultSamplingProbability
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: call floatEquals for the right side?

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 16]
```diff
@@ -0,0 +1,513 @@
...
+		}
+	}
+	return false
+}
+
+// generateStrategyResponses generates a SamplingStrategyResponse from the calculated sampling probabilities.
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: generates and caches

File: plugin/sampling/strategystore/adaptive/weightscache.go [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+
+package adaptive
+
+import (
+	"math"
+	"sync"
+)
+
+// weightsCache stores normalizing weights of different lengths. The head of the weights slice
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/weightscache.go]: weight vectors of different lengths. Could've used 'vector' in the type name as well, makes it much clearer.

File: plugin/sampling/strategystore/adaptive/weightscache.go [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+// getWeights returns normalizing weights for the specified length.
+func (c *weightsCache) getWeights(length int) []float64 {
+	c.Lock()
+	defer c.Unlock()
+	if weights, ok := c.cache[length]; ok {
+		return weights
+	}
+	l := float64(length)
+	// closed form of sum l^4 ie 1^4 + 2^4 + ... + l^4
+	sum := (l / 30) * (l + 1) * ((2 * l) + 1) * ((3 * l * l) + (3 * l) - 1)
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/weightscache.go]: I have no idea what's going on here

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 19]
```diff
@@ -0,0 +1,522 @@
...
+	}
+	var newProbability float64
+	if floatEquals(qps, 0) {
+		// Edge case; we double the sampling probability if the QPS is 0 so that we force the service
+		// to at least sample one span probabilistically.
+		newProbability = oldProbability * 2.0
+	} else {
+		newProbability = p.probabilityCalculator.Calculate(p.TargetSamplesPerSecond, qps, oldProbability)
+	}
+	return math.Min(maxSamplingProbability, math.Max(p.MinSamplingProbability, newProbability))
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/adaptive/processor.go]: the `maxSamplingProbability` const is superfluous, this stmt would read better if you simply use `1.0`

**Comment by black-adder** [comment]
going to land and address comments and wire things up in next PR

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 20]
```diff
@@ -0,0 +1,506 @@
...
+
+const (
+	maxSamplingProbability = 1.0
+
+	getThroughputErrMsg = "failed to get throughput from storage"
+
+	defaultFollowerProbabilityInterval = 20 * time.Second
+
+	// The number of past entries for samplingCache the leader keeps in memory
+	serviceCacheSize = 25
```
[the remark this answers is not in the document]
Comment (black-adder) [reply plugin/sampling/strategystore/adaptive/processor.go]: yes we only use the most recent two entries in the serviceCache. We can probably remove the older entries in the cache. I'll add a TODO for now because there might've been a reason my younger and smarter self did this.

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 21]
```diff
@@ -0,0 +1,506 @@
...
+}
+
+// Start initializes and starts the sampling processor which regularly calculates sampling probabilities.
+func (p *processor) Start() error {
+	p.logger.Info("starting sampling processor")
+	p.shutdown = make(chan struct{})
+	if starter, ok := p.electionParticipant.(jio.Starter); ok {
+		starter.Start()
+	}
+	p.loadProbabilities()
```
[the remark this answers is not in the document]
Comment (black-adder) [reply plugin/sampling/strategystore/adaptive/processor.go]: - master: since there's no probability to compare against to determine if an operation is using adaptive sampling, the processor will start calculating probabilities again from the default prob 0.001.
- follower: will return default probabilities to agents/clients.

File: plugin/sampling/strategystore/adaptive/processor.go [hunk 22]
```diff
@@ -0,0 +1,522 @@
...
+	p.throughputs = append([]*throughputBucket{bucket}, p.throughputs...)
+	if len(p.throughputs) > p.AggregationBuckets {
+		p.throughputs = p.throughputs[0:p.AggregationBuckets]
+	}
+}
+
+// aggregateThroughput aggregates operation throughput from different buckets into one.
+// All input buckets represent a single time range, but there are many of them because
+// they are all independently generated by different collector instances from inbound span traffic.
+func (p *processor) aggregateThroughput(throughputs []*model.Throughput) serviceOperationThroughput {
```
[the remark this answers is not in the document]
Comment (black-adder) [reply plugin/sampling/strategystore/adaptive/processor.go]: See, the code isn't _that_ hard to follow.

File: plugin/sampling/strategystore/adaptive/factory.go [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+// Initialize implements strategystore.Factory
+func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+	f.logger = logger
+	f.metricsFactory = metricsFactory
+	return nil
+}
+
+// CreateStrategyStore implements strategystore.Factory
+func (f *Factory) CreateStrategyStore() (strategystore.StrategyStore, error) {
+	// TODO
```
[the remark this answers is not in the document]
Comment (Ashmita152) [reply plugin/sampling/strategystore/adaptive/factory.go:61]: @yurishkuro I can see that in processor.go, the `NewProcessor()` (https://github.com/jaegertracing/jaeger/blob/master/plugin/sampling/strategystore/adaptive/processor.go#L119-L152) is already returning a `&processor{}` that has implemented the `GetSamplingStrategy`(https://github.com/jaegertracing/jaeger/blob/master/plugin/sampling/strategystore/adaptive/processor.go#L155-L162) and satisfies the `StrategyStore` interface. We no longer use the Factory interface anywhere inside the adaptive sampling directory (https://github.com/jaegertracing/jaeger/tree/master/plugin/sampling/strategystore/adaptive). Do we need an implementation for Factory{} inside adaptive package or can we proceed without it?

[the remark this answers is not in the document]
Comment (yurishkuro) [reply plugin/sampling/strategystore/adaptive/factory.go:61]: individual factories are used by the meta-factory `plugin/sampling/strategystore/factory.go`:

https://github.com/jaegertracing/jaeger/blob/5062366157fff65a052dc6ac6779e7621054248d/plugin/sampling/strategystore/factory.go#L34
